File: String/levenshtein_distance.py
# Levenshtein distance

def lev_dist(a:str, b:str)->int:
    m = len(a)
    n = len(b)
    d = [[None]*(n+1) for i in range(m+1)]

    for i in range(m+1):
        for j in range(n+1):
            if(i == 0):
                d[i][j]=j
            elif(j == 0):
                d[i][j]=i
            elif(a[i-1] == b[j-1]):
                d[i][j] = d[i-1][j-1]
            else:
                d[i][j] = min(min(1 + d[i-1][j], 1 +d[i][j-1]), 1 + d[i-1][j-1])
    return d[m][n]


if __name__=='__main__':
    a = "algorithm"
    b = "algebraic"
    print(lev_dist(a,b)) # 6

File: String/LevenshteinDistance.cpp
/*Levenshtein Distance*/

#include <bits/stdc++.h>
using namespace std;

#define fast_io ios::sync_with_stdio(false)

int levenshteinDistance(const string &s1, const string &s2) {
	int m = s1.length();
	int n = s2.length();
	int dp[m+1][n+1];
	for(int i = 0; i <= m; ++i) {
		dp[i][0]=i;
	}
	for(int j = 0; j <= n; ++j) {
		dp[0][j]=j;
	}
	for(int i = 1; i <= m; ++i) {
		for(int j = 1; j <= n; ++j) {
			dp[i][j]=min({dp[i][j-1]+1,dp[i-1][j]+1,dp[i-1][j-1]+(s1[i-1]==s2[j-1] ? 0:1)});
		}
	}
	return dp[m][n];
}

int main() {
	fast_io;
	cin.tie(0);
	string s; cin >> s;
	string t; cin >> t;
	cout << levenshteinDistance(s,t);
	return 0;
}
